src/librbd/journal/RemoveRequest.cc

```diff
@@ -0,0 +1,155 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "common/dout.h"
+#include "common/errno.h"
+#include "common/Timer.h"
+#include "common/WorkQueue.h"
+#include "include/assert.h"
+#include "librbd/Utils.h"
+#include "librbd/journal/RemoveRequest.h"
+
+#define dout_subsys ceph_subsys_rbd
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::Journal::RemoveRequest: "
+
+namespace librbd {
+
+using util::create_context_callback;
+
+namespace journal {
+
+template<typename I>
+RemoveRequest<I>::RemoveRequest(IoCtx &ioctx, const std::string &image_id,
+                                              const std::string &client_id,
+                                              ContextWQ *op_work_queue,
+                                              Context *on_finish)
+  : m_image_id(image_id), m_image_client_id(client_id),
+    m_op_work_queue(op_work_queue), m_on_finish(on_finish) {
+  m_ioctx.dup(ioctx);
+  m_cct = reinterpret_cast<CephContext *>(m_ioctx.cct());
+}
+
+template<typename I>
+void RemoveRequest<I>::send() {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  init_journaler();
+}
+
+template<typename I>
+void RemoveRequest<I>::init_journaler() {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  ImageCtx::get_timer_instance(m_cct, &m_timer, &m_timer_lock);
+  m_journaler = new Journaler(m_op_work_queue, m_timer, m_timer_lock,
+                              m_ioctx, m_image_id, m_image_client_id,
+                              m_cct->_conf->rbd_journal_commit_age);
+
+  using klass = RemoveRequest<I>;
+  Context *ctx = create_context_callback<klass, &klass::handle_init_journaler>(this);
+
+  m_journaler->init(ctx);
```

Comment:
-ENOENT can also be returned if the client isn't registered. If you stat it first and get -ENOENT, you can silently exit the state machine. Otherwise, you can continue to init and remove.